src/ale_interface.cpp

```diff
@@ -269,31 +269,32 @@ reward_t ALEInterface::act(Action action) {
 // Returns the vector of modes available for the current game.
 // This should be called only after the rom is loaded.
 ModeVect ALEInterface::getAvailableModes() {
-  throw std::runtime_error("This feature is not available yet.");
-  ModeVect placeholder;
-  return placeholder;
+  return romSettings->getAvailableModes();
 }
 
 // Sets the mode of the game.
 // The mode must be an available mode.
 // This should be called only after the rom is loaded.
 void ALEInterface::setMode(game_mode_t m) {
-  throw std::runtime_error("This feature is not available yet.");
+  romSettings->setMode(m, theOSystem->console().system(), *(environment.get()));
 }
 
 //Returns the vector of difficulties available for the current game.
 //This should be called only after the rom is loaded.
 DifficultyVect ALEInterface::getAvailableDifficulties() {
-  throw std::runtime_error("This feature is not available yet.");
-  DifficultyVect placeholder;
-  return placeholder;
+  return romSettings->getAvailableDifficulties();
 }
 
 // Sets the difficulty of the game.
 // The difficulty must be an available mode.
 // This should be called only after the rom is loaded.
 void ALEInterface::setDifficulty(difficulty_t m) {
-  throw std::runtime_error("This feature is not available yet.");
+  DifficultyVect available = romSettings->getAvailableDifficulties();
```

Comment:
Is "difficulty" really a single value? Is this how we want to think of it?

Reply:
I'm not sure I understand your question here. There is a right and a left difficulty switch in the console. Therefore, 4 possible difficulties (2 bits). What do you mean by "want to think of it"?

Reply:
It seems more natural to me to think of the difficulty settings as "Player 1 Difficulty" and "Player 2 Difficulty", rather than 4 separate choices. Let me ask a different question first: when is "available difficulties" anything else than the (A, A), (A, B), (B, A), (B, B) ?

Reply:
Marc, there are 4 difficulties. They are not tied to a specific player. For example, Venture, has 4 difficulties ([manual](https://atariage.com/manual_html_page.php?SoftwareLabelID=576)). Of course, if you want, I can change the mask to be something else, but I just want to understand what you have in mind first.

```
   Skill          Switch
   Level          Setting
     1         left B/right B
     2         left B/right A
     3         left A/right B
     4         left A/right A
```

Reply:
I see, that makes a lot more sense then. Do you want to put this table somewhere as a comment?

Reply:
Ok. I'll do that. But I'll write it as an illustrative example because, as I said, this table is from `Venture`.